### PR TITLE
First Draft: Push/Stream notifications

### PR DESCRIPTION
**Pull request by dsanghan**
Known issues:
- Need to change protocol to allow streaming responses for streaming subscription get events. Right now, requests only returns data once a request is fully finished (i.e when there's a timeout for the streaming endpoint), but needs to support partial data responses.

Not sure how to fix the above, so thought it best to start a discussion.

### CODE REVIEW

**Comment by dsanghan** [comment]
Yup. Might take some time for that. 

I've also added SyncFolderItems - again first draft, would love feedback / fixes.

**Comment by ecederstrand** [comment]
This looks really good. Looks ready to merge. I'd just like some documentation or tests, so I can test it out myself before merging. A quick POC here in the comments is fine as a start.

**Comment by ecederstrand** [comment]
Refs #145 and #209

**Comment by dsanghan** [comment]
@ecederstrand: I need to finish the remaining todo before this is ready for test/merge. I'll try and get to it soon. I added some docs to make the methods easier to understand as well.

**Comment by ecederstrand** [comment]
No problem. Keep up the  good work :-)

**Comment by ecederstrand** [comment]
Hi @dsanghan I'm preparing to roll a new release, since it's been a long time. Do you need more time to complete this PR, or can you get it done within the next couple of days?

Feel free to speak up if you need help or guidance :-)

**Comment by RemiDesgrange** [comment]
Is there anything I can do to help ? I'm very interesting in this feature.

File: exchangelib/items.py [hunk 1]
```diff
@@ -1317,5 +1317,38 @@ def __hash__(self):
...
+    FIELDS = Item.FIELDS
+
+
+class CreateChange(Item):
+    """
+    MSDN: https://msdn.microsoft.com/en-us/library/office/aa565609(v=exchg.150).aspx
+    """
+    ELEMENT_NAME = 'Create'
+    FIELDS = Item.FIELDS + [
+        EWSElementField('changed_item', field_uri='Message', value_cls=Message)
```
Comment (stevedh) [inline exchangelib/items.py]: FYI I believe you need to add all the fields here (https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/create-itemsync) to handle other element types.  I didn't check the other parser classes, but they also look a little think.

Thanks for the patch!

**Comment by mandarvaze** [comment]
@ecederstrand Can you provide *tentative* date for when this PR might get merged ?

**Comment by ecederstrand** [comment]
@mandarvaze and @spmsh 

This is a contribution from @dsanghan. I'm not actively working on it and have no immediate plans to do so.

This PR is not ready to be merged yet. See https://github.com/ecederstrand/exchangelib/pull/355#issuecomment-393547577

**Comment by mandarvaze** [comment]
@spmsh @dsanghan  Can either of you provide some documentation about how you tested this ? As @ecederstrand suggested test cases or documentation (like in the README) would be helpful.

In the meantime, I'm looking at the code, and trying to understand the usage pattern. I require notifications mainly for NewEmails

**Comment by spmsh** [comment]
Hello ~

I've made a quick and probably dirty update to the PR to get the GetStreamNotifications happening in real time rather than waiting for the timeout to get the whole lot of notifications (i.e., what is referenced as a known in the very first post of this thread).

Added stream=True in subscribe.py > `GetStreamingEvents` class > `call` function:

    class GetStreamingEvents(EWSAccountService):
        """
        https://msdn.microsoft.com/en-us/library/office/dn458792(v=exchg.150).aspx
        """
        SERVICE_NAME = 'GetStreamingEvents'
        element_container_name = '{%s}Notifications' % MNS
    
        def call(self, subscription_ids):
            self.streaming = True 
            return self._get_elements(payload=self.get_payload(subscription_ids))

Then in util.py > `post_ratelimited` function:

    def post_ratelimited(protocol, session, url, headers, data, allow_redirects=False, stream=False):
        [...] # some code
            while True:
                _back_off_if_needed(protocol.credentials.back_off_until)
                log.debug('Session %s thread %s: retry %s timeout %s POST\'ing to %s after %ss wait', session.session_id,
                          thread_id, retry, protocol.TIMEOUT, url, wait)
                d_start = time_func()
                # Always create a dummy response for logging purposes, in case we fail in the following
                r = DummyResponse(url=url, headers={}, request_headers=headers)
                try:
                    r = session.post(url=url, headers=headers, data=data, allow_redirects=False, timeout=protocol.TIMEOUT,
                                     stream=stream)
                    # modification are below here
                    if stream: # stream was set as True for the GetStreamingEvents call
                        log.info("Streaming is on ...", r.status_code)
                        chunk_size = 512 * 1024
                        notification = bytearray()
                        for line in r.iter_content(chunk_size=chunk_size):
                            if len(line) != 1:
                                notification.extend(line)
                            else:
                                if len(notification) > 0: # empty notification is sent by EWS to delimit between notifications
                                    do_something_with_the_notification(notification) # Microsoft recommends to manage this in another thread
        [...] # rest of code unmodified

With this, the connection stays open until the timeout, but keeps streaming notifications as soon as something is happening. You get the notification and need to manage it in a different thread.

I'm not a pro at clean code, you might incorporate this better in the whole lib, but I confirm you guys that this is working as intended and notifications are receiving in real time. At the end of the timeout, just need to re-open the connection.

~spmsh

**Comment by lululukas** [comment]
Hi Erik,

I would like to contribute to this, I have implemented most of the changes suggested in your review and @spmsh changes. Since this PR was opened more than a year ago there are probably quite a few more changes that have to be done. I don't think I can push to this PR, since it's on @dsanghan 's fork, or can I? Should I open a new PR then?

**Comment by ecederstrand** [comment]
@lululukas You can't commit to this PR, but you can fork @dsanghan's fork and create a new PR against exchangelib.

**Comment by dsanghan** [comment]
@lululukas @ecederstrand Might be best to create a new PR and I'll just delete this one.

**Comment by ecederstrand** [comment]
Oh, and just in case you are wondering: I'm not opposed to the work being done here :smile: 

It just needs an agreement on the public API, the API needs to be documented in a new section in the README, and it needs basic test cases exercising the API.

**Comment by sammerpuranpost** [comment]
Has there been any work done here since the last time? Asking because such a feature would be handy for me

**Comment by ecederstrand** [comment]
No, there has been no work on this, AFAIK.

**Comment by ecederstrand** [comment]
Push and stream notification support was added in a series of commits ending in a729eac

**Pull request closed without merging**